find_exponential/src/main.rs

```diff
@@ -0,0 +1,270 @@
+            println!("{:<20} throughput: {}", count, count as u64 / start_time.elapsed().as_secs());
+        }
+    }
+}
+
+fn test(pattern: &str) -> f64 {
+    loop {
+        let res1 = test_pattern(pattern, SAMPLE_SIZE, false);
+        if let PatternResult::Linear(_array, score) = res1 {
+            return score;
+        }
+        let res2 = test_pattern(pattern, SAMPLE_SIZE*2, true);
+        if let PatternResult::Linear(_array, score) = res2 {
+            return score;
+        }
+
+        // possible non-linear behaviour found
+
+        let score = match res2 {
+            PatternResult::Linear(..) => unreachable!(),
+            PatternResult::NonLinear(array, score) => {
+                println!(
+                    "\n\
+                    possible non-linear behaviour found\n\
+                    pattern: {:?}\n\
+                    score: {}\n\
+                    {:?}\n",
+                    pattern,
+                    score,
+                    array,
+                );
+                score
+            },
+            PatternResult::TooLong(array) => {
+                println!(
+                    "\n\
+                    possible non-linear behaviour found due to exceeding MAX_MILLIS (parsing took too long)\n\
+                    pattern: {:?}\n\
+                    score: 0\n\
+                    {:?}\n",
+                    pattern,
+                    array,
+                );
+                0.0
+            },
+            PatternResult::HugeOutlier(array, trim_len) => {
+                println!(
+                    "\n\
+                    huge outlier found, this may indicate weird behaviour in pulldown-cmark\n\
+                    pattern: {:?}\n\
+                    score: 0\n\
+                    trim length: {}\n\
+                    trimmed rest: {:?}\n\
+                    {:?}\n",
+                    pattern,
+                    trim_len,
+                    &pattern[..pattern.len() - trim_len],
+                    array,
+                );
+                0.0
+            },
+        };
+        return score;
+    }
+}
+
+#[derive(Debug)]
+enum PatternResult {
+    /// Probably linear growth (points, score)
+    Linear(Vec<[f64; 2]>, f64),
+    /// Probably non-linear growth (points, score)
+    NonLinear(Vec<[f64; 2]>, f64),
+    /// Execution took too long, assumed non-linear
+    TooLong(Vec<[f64; 2]>),
+    /// Very slow outlier, probably due to different parsing in pulldown-cmark.
+    /// (points, number of bytes truncated from end reaching this behaviour)
+    ///
+    /// Possibly indicates weird behaviour in pulldown-cmark.
+    /// See <https://github.com/raphlinus/pulldown-cmark/issues/287> for an example.
+    HugeOutlier(Vec<[f64; 2]>, usize),
+}
+
+fn test_pattern(pattern: &str, sample_size: usize, recalculate_outliers: bool) -> PatternResult {
+    let s = pattern.repeat(NUM_BYTES / pattern.len());
+    let mut array = Vec::with_capacity(sample_size);
```

Comment:
Make the caller pass `&mut [[f64; 2]; 10][..sample_size]` instead of using a `Vec`, no allocations required. Only allocate in `pearson_correlation`.

Reply:
Done in ca8e59e94b074039cf42fe36de5563f33ae5a7c5. Didn't seem to boost the throughput.